CFE-4070: Added commands `cfbs get-input` and `cfbs set-input`

larsewi opened the pull request.

larsewi committed:
Added new cfbs set-input command

Ticket: CFE-4070
Changelog: Title
Co-authored-by: Ole Herman Schumacher Elgesem <olehermanse@users.noreply.github.com>
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Added new cfbs get-input command

Ticket: CFE-4070
Changelog: Title
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Added shell test for cfbs get-input/set-input

Ticket: 4070
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

cf-bottom commented:
Thanks for submitting a pull request! Maybe @olehermanse can review this?

olehermanse commented on cfbs/commands.py:
```diff
@@ -1000,3 +1001,103 @@ def input_command(args, input_from="cfbs input"):
+
+    if src_path is None:
+        try:
+            data = json.load(stdin)
```
Use `OrderedDict` explicitly so python 3.5 also has predictable dictionaries:

```suggestion
            data = json.load(stdin, object_pairs_hook=OrderedDict)
```

olehermanse commented on cfbs/commands.py:
```diff
@@ -7,6 +7,7 @@
 import copy
 import logging as log
 import json
```
```suggestion
import json
from collections import OrderedDict
```

olehermanse reviewed (changes requested):
Python 3.5 not working as 3.6 is the first one were dictionaries are always oredered.

olehermanse commented on tests/shell/030_get_set_input.sh:
```diff
@@ -0,0 +1,99 @@
+]' > expected.output
+diff actual.output expected.output
+
+# Igor changes the wrong field
+echo '[
+  {
+    "type": "string",
+    "variable": "bogus",
+    "label": "Filename",
+    "question": "What file should this module create?",
+    "response": "/tmp/test-2.txt"
+  }
+]' > igors-input.json
+
+! cfbs set-input create-single-file igors-input.json
```
Could you add another test like this where one of the keys is changed, like `type` is changed to something else?

larsewi replied:
Sure thing! That a key is changed instead of value?

olehermanse replied:
yep, thanks.

vpodzime commented on cfbs/commands.py:
```diff
@@ -1000,3 +1001,74 @@ def input_command(args, input_from="cfbs input"):
+
+    if src_path is None:
+        try:
+            data = json.load(stdin)
```
If the user runs the command `cfbs set-input some-module`, they will get a seemingly stuck/hanging tool. We should either add some prompt or change the stdin-based input to be based on an extra option or a special value of the argument (`-` is typically used for this).

larsewi replied:
Made the second argument required, and added support for the special value `-`, which is `stdin`/`stdout` for get-input/set-input respectively

vpodzime commented on cfbs/commands.py:
```diff
@@ -1000,3 +1001,74 @@ def input_command(args, input_from="cfbs input"):
+        if set(a.keys()) != set(b.keys()) - ignore:
+            return False
+        """Avoid code duplication by converting the values of the two dicts
+        into two lists in the same order and compare the lists instead"""
```
Please use comments for what is commenting the code. Documentation strings are for documentation.

vpodzime commented on cfbs/main.py:
```diff
@@ -119,6 +119,11 @@ def main() -> int:
+    if args.command == "set-input":
+        if len(args.args) < 1:
+            user_error("Missing argument <module-name>")
+        module, file = args.args[0], args.args[1] if len(args.args) >= 2 else None
```
Please use parentheses to make this easier to read.

vpodzime commented on cfbs/commands.py:
```diff
@@ -1000,3 +1001,74 @@ def input_command(args, input_from="cfbs input"):
+        keys = a.keys()
+        return _compare_list([a[key] for key in keys], [b[key] for key in keys])
+
+    def _compare_list(a, b):
```
So this requires that things are in the very same order. I'm not sure how much strict JSON is regarding ordering of object's children, but it feels wrong to me to require the inputs to be in the very same order.

larsewi replied:
Objects does not have to be in order. I added a test to prove that. But lists do, and they should

vpodzime replied:
Aren't the inputs in a list?

larsewi replied:
Yep! Imagine the following variables being swapped
```
$ cfbs input create-single-file-with-content
--snip--
What file do you want to create? /tmp/testfile
What content should this file have? Hello CFEngine!
```
i.e.
```
$ cfbs input create-single-file-with-content
--snip--
What content should this file have? Hello CFEngine!
What file do you want to create? /tmp/testfile
```
it would not make much sense for the user.

vpodzime replied:
It wouldn't. But when the data comes from a form in MP or generally from some JSON, ordering should not matter.

olehermanse replied:
The order of the list defines the order of the prompts and the order of the UI elements in MP. It matters. MP should be working on the list it got from get-input, and should not reorder the elements before running set-input.

olehermanse replied:
We could enable module authors to more easily change input (reorder it) and have it work automatically with `cfbs update`, but that's not this command.

vpodzime replied:
> should not reorder the elements before running set-input
I don't see a good reason for this to be disallowed/unsupported.

olehermanse replied:
The order here communicates how it should be shown to the user and when it comes back, how it was shown to the user.

For example, things coming back in another order could indicate that MP is not showing things correctly, or not handling the input received from the user correctly.

It is important to record what the user saw and what they responded, including the order of things. @larsewi showed one example. Similar situations can apply to the GUI. If labels are bad or confusing, or users don't read them, or module authors made mistakes when filling out the labels or variable names, having an accurate record of what the user saw and entered is valuable.

I would take it one step further and say we should also record whether the user saw the prompt (used the CLI) or whether they saw the label (used MP). However this is a bit more difficult, and we are not doing that currently.

vpodzime replied:
I guess I'm thinking about this differently -- the input, to me, is just data. So I would expect the `get-input` command to only take answers, basically as `input_id: value` pairs

vpodzime commented on cfbs/commands.py:
```diff
@@ -1072,3 +1072,32 @@ def _compare_list(a, b):
+        print(pretty(data))
+    else:
+        log.debug("Writing json to '%s'" % dst_path)
+        write_json(dst_path, data)
```
I would expect `get-input module some-file` to extract the input from `some-file`. Again, I'd prefer an extra option for the output.

vpodzime reviewed (commented):
Looks OK to me otherwise.

olehermanse commented on cfbs/main.py:
```diff
@@ -119,6 +119,13 @@ def main() -> int:
     if args.command == "input":
         return commands.input_command(args.args)
+    if args.command in ("set-input", "get-input"):
+        if len(args.args) < 1:
+            user_error("Missing argument <module-name>")
+        module, file = args.args[0], (args.args[1] if len(args.args) >= 2 else None)
+        if args.command == "set-input":
+            return commands.set_input_command(module, file)
+        return commands.get_input_command(module, file)
```
As @vpodzime pointed out, you can consider adding a requirement to specify filename or `-` for stdin here;

```
cfbs get-input <module-name> <filename or - for stdin>
cfbs set-input <module-name> <filename or - for stdin>
```

Then you can always require `len(args.args) == 2`, and print out a helpful error message:

```
error: Filename required as last argument (or - for stdin)
```

I don't think we should make it more complicated than this by using `--`-options or prompts.

olehermanse reviewed (approved):
LGTM. There are some comments above which you can read / considers, but no blockers from my side.

larsewi committed:
cfbs get-input also looks for module in index

cfbs get-input now also looks for module in index if module not found in build.

Ticket: None
Changelog: Body
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

olehermanse commented on cfbs/main.py:
```diff
@@ -119,6 +120,34 @@ def main() -> int:
         return commands.input_command(args.args)
+    if args.command == "set-input":
+        if len(args.args) != 2:
+            log.error("Filename required as argument (or - for stdin)")
```
```suggestion
            if len(args.args) == 1:
                log.error("Filename required as argument (or - for stdin)")
            if len(args.args) == 0:
                log.error("Module and filename (or - for stdin) must be specified")
            if len(args.args) > 2:
                log.error("Too many arguments, set-input only accepts 2 (module and filename)")
```

olehermanse commented on cfbs/main.py:
```diff
@@ -119,6 +120,34 @@ def main() -> int:
+            infile.close()
+    if args.command == "get-input":
+        if len(args.args) != 2:
+            log.error("Filename required as argument (or - for stdout)")
```
Same here.

olehermanse reviewed (changes requested):
Almost ;)

olehermanse reviewed (approved):
LGTM now
